src/lm/arpa-sampling.h

```diff
@@ -0,0 +1,129 @@
+// arpa_sampling.h
+
+// Copyright     2016  Ke Li
+
+// See ../COPYING for clarification regarding multiple authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
+// WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABILITY OR NON-INFRINGEMENT.
+// See the Apache 2 License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef ARPA_SAMPLING_H_
+#define ARPA_SAMPLING_H_
+
+#include <algorithm>
+#include <map>
+#include <iostream>
+#include <string>
+#include <sstream>
+#include <fstream>
+#include <vector>
+
+#include <sys/time.h>
+#include <unistd.h>
+#include <stdlib.h>
+#include <math.h>
+#include "fst/fstlib.h"
+#include "util/common-utils.h"
+#include "arpa-file-parser.h"
+
+namespace kaldi {
+
+class ArpaSampling : public ArpaFileParser {
+ public:
+  friend class ArpaSamplingTest;
+
+  // HistType represents a history
+  typedef std::vector<int32> HistType;
+  // WordToProbsMap represents the words and their probabilities given
+  // an arbitrary history
+  typedef std::unordered_map<int32, std::pair<BaseFloat, BaseFloat> > WordToProbsMap; 
+  // NgramType represents the map between a history and the map of existing words
+  // and their probabilities given this history
+  typedef std::unordered_map<HistType, WordToProbsMap, VectorHasher<int32> > NgramType;
+  // HistWeightsType represents the map between histories (with weights) and
+  // their computed weights
+  typedef std::unordered_map<HistType, BaseFloat, VectorHasher<int32> > HistWeightsType;
+
+  // constructor
+  ArpaSampling(ArpaParseOptions options, fst::SymbolTable* symbols)
+     : ArpaFileParser(options, symbols) { 
+       ngram_order_ = 0;
+       num_words_ = 0;
+       bos_symbol_ = "<s>";
+       eos_symbol_ = "</s>";
+       unk_symbol_ = "<unk>";
+  }
+  
+  // This function computes the unigram distribution of all vocab words which are
+  // represented by integers from 1 to num_words (0 is reserved for epsilon)
+  // Probabilities of integers are not represented by any valid word is 0, e.g.
+  // 0 has probability 0.0
+  void GetUnigramDistribution(std::vector<BaseFloat> *unigram_probs);
```

Comment:
I'll look more into it later today or maybe tomorrow, but my immediate concern is the assumptions that which this interface makes about symbols. OpenFST assumes that the *states* are dense, in the sense that if a state s_i exists, then s_j exists for any j<i. AFAIK, no such assumption is ever made about symbol indices. Generally, numbering of symbols allows arbitrary gaps. The interface (both this and the next functions) assumes that symbols are numbered 1, ..., N_1, where N_1 is the number of unigrams in the model, i. e. that the set of unigrams read from the model is just a permutation of the range [1..N_1]. This would not hold if the symbol map (normally our `words.txt`) contains numbering gaps (which is not normally the case but was not prohibited until now). This would also not hold if not all words are present in the LM (which might often be the case!) For example of the latter, if the file `words.txt` defines words A=1, B=2, C=3 and D=4, but only C and D are in the LM, this function is set to return the vector of the length 3, with the probs for symbols 0, 1, and 2, which are all 0.

I think it would be more consistent to make no assumptions about the values of the symbol IDs, and rather return a partial mapping of symbols to the probs (unordered_map<int32, BaseFloat>), or provide a method to return a probability given symbol ID. I do not know about the intended use for this class to make a sensible suggestion here, though.